RFS/src/main/java/org/opensearch/migrations/bulkload/common/DocumentReindexer.java

```diff
@@ -19,26 +23,53 @@
 import reactor.core.scheduler.Schedulers;
 
 @Slf4j
-@RequiredArgsConstructor
 public class DocumentReindexer {
+    private static final Supplier<IJsonTransformer> NOOP_TRANSFORMER_SUPPLIER = () -> new NoopTransformerProvider().createTransformer(null);
 
     protected final OpenSearchClient client;
     private final int maxDocsPerBulkRequest;
     private final long maxBytesPerBulkRequest;
     private final int maxConcurrentWorkItems;
-    private final Supplier<IJsonTransformer> transformerSupplier;
+    private final ThreadSafeTransformerWrapper threadSafeTransformer;
+    private final boolean isNoopTransformer;
+
+    public DocumentReindexer(OpenSearchClient client,
+               int maxDocsPerBulkRequest,
+               long maxBytesPerBulkRequest,
+               int maxConcurrentWorkItems,
+               Supplier<IJsonTransformer> transformerSupplier) {
+        this.client = client;
+        this.maxDocsPerBulkRequest = maxDocsPerBulkRequest;
+        this.maxBytesPerBulkRequest = maxBytesPerBulkRequest;
+        this.maxConcurrentWorkItems = maxConcurrentWorkItems;
+        this.isNoopTransformer = transformerSupplier == null;
+        this.threadSafeTransformer = new ThreadSafeTransformerWrapper((this.isNoopTransformer) ? NOOP_TRANSFORMER_SUPPLIER : transformerSupplier);
```

Comment:
can you shake the isNoopTransformer variable and just make the threadSafeTransformer be the Noop transformer, then your virtual call does the identity function job w/ the cost of a virtual call (but simpler code).